op-node/rollup/derive/engine_queue.go

```diff
@@ -237,12 +240,22 @@ func (eq *EngineQueue) LowestQueuedUnsafeBlock() eth.L2BlockRef {
 	return ref
 }
 
+func (eq *EngineQueue) BackupUnsafeL2Head() eth.L2BlockRef {
+	return eq.ec.BackupUnsafeL2Head()
+}
+
 // Determine if the engine is syncing to the target block
 func (eq *EngineQueue) isEngineSyncing() bool {
 	return eq.ec.IsEngineSyncing()
 }
 
 func (eq *EngineQueue) Step(ctx context.Context) error {
+	// If we don't need to call FCU to restore unsafeHead using backupUnsafe, keep going b/c
+	// this was a no-op(except correcting invalid state when backupUnsafe is empty but reorg triggered).
+	// If we needed to perform a network call, then we should yield even if we did not encounter an error.
+	if err := eq.ec.TryBackupUnsafeReorg(ctx); !errors.Is(err, errNoBackupUnsafeReorgNeeded) {
+		return err
```

Comment:
I had a little bit of trouble understanding the behavior of `TryBackupUnsafeReorg` and at first assumed that this error check had meant to also check `!=nil`. The comments above helped me realize it's intentional, but it reveals that maybe this function should be split into two parts. As it is today, there are three possible outcomes of `err`:
- `nil` meaning the work was done and nothing went wrong
- `errNoBackupUnsafeReorgNeeded` meaning that no work was done
- `anything else` meaning work was done and something went wrong

I would argue that `errNoBackupUnsafeReorgNeeded` is a flag/signal, not an error. Seeing as we run this function on every `Step`, having the function return an error indicating "*you did not need to run me*" seems counterintuitive. Presumably it's going to return that error the vast majority of the time.

If it were instead split into two parts, you'd have a check, `shouldTryBackupUnsafeReorg()` return a boolean,
- `true` if the conditions to call the function are met (`needed` flag, not-syncing, and not-empty)
- `false` otherwise

And then if it's true, you can simply `return TryBackupUnsafeReorg` because you know you'll be consuming resources. This would let you eliminate the entire `errNoBackupUnsafeReorgNeeded` error type, and would minimize calls to the the actual `TryBackupUnsafeReorg` function to just instances where it is needed.

Reply:
Another possibility, if you don't want to split this into two functions: have `TryBackupUnsafeReorg` return `(bool, error)`. The bool as a signal that work was done, and then the error is related to the result of that work. You'd use it the same way as when checking a map for a value,

```
if ran, err := eq.ec.TryBackupUnsafeReorg(ctx); ran {
    return err
}
```

Reply:
Fixed at https://github.com/ethereum-optimism/optimism/pull/8925/commits/c4778d8eec00917b601662066ec3e9da8a76e7fa.

I mixed your both suggestions; modularizing out checks at `shouldTryBackupUnsafeReorg`, and returning `(bool, error)` at `TryBackupUnsafeReorg`. Did not wanted to expose `shouldTryBackupUnsafeReorg` outside engine controller(leaking abstraction). The refactoring resulted in eliminating `errNoBackupUnsafeReorgNeeded` which is nice.

After refactoring, I feel like similar refactoring approach can be applied to `TryUpdateEngine` as well, resulting in removing `errNoFCUNeeded`. What do you think?

Reply:
Could do that, yeah! Let's not extend the refactor past the new code for now, thank you for updating this.